File: packages/r/R/impact.R
#' Run the Inclusive Multimorbidity Phenotyping Algorithm and Coding Tool.
#'
#' Each row of `data` is a coded event. IMPACT returns that row's
#' identifier and either 321 granular LTC indicators or 116 grouped phenotype
#' indicators. The output level must be selected explicitly.
#'
#' @param data A data frame containing coded events.
#' @param id Name of the identifier column. Names beginning `__` are reserved
#'   for IMPACT outputs.
#' @param codesystems Code system name(s), in the same order as
#'   `searchvars`.
#' @param searchvars A list with one character vector of character code-column
#'   names per code system. For one code system, a character vector is also
#'   accepted.
#' @param level Either "ltc" or "phenotype".
#' @param multimorbidity A single logical value. If TRUE, add event-level,
#'   phenotype-based counts.
#' @param summary A single logical value. If TRUE, print matched-code counts by
#'   code system.
#' @return A data frame containing the identifier and requested indicators.
#' @export
impact <- function(data, id, codesystems, searchvars, level,
                   multimorbidity = FALSE, summary = FALSE) {
  if (!is.data.frame(data)) stop("data must be a data frame.", call. = FALSE)
  if (!is.character(id) || length(id) != 1L || is.na(id)) {
    stop("id must be one column name.", call. = FALSE)
  }
  if (startsWith(id, "__")) {
    stop(
      "Identifier column names beginning '__' are reserved for IMPACT outputs.",
      call. = FALSE
    )
  }
  if (!(id %in% names(data))) {
    stop("Identifier column '", id, "' was not found.", call. = FALSE)
  }
  .impact_validate_logical(multimorbidity, "multimorbidity")
  .impact_validate_logical(summary, "summary")
  if (missing(level) || length(level) != 1L) {
    stop("level must be specified as 'ltc' or 'phenotype'.", call. = FALSE)
  }
  level <- match.arg(tolower(level), c("ltc", "phenotype"))
  codesystems <- .impact_as_codesystems(codesystems)
  if (!is.list(searchvars)) {
    searchvars <- if (length(codesystems) == 1L) list(searchvars) else as.list(searchvars)
  }
  searchvars <- lapply(searchvars, function(x) as.character(unlist(x, use.names = FALSE)))
  if (length(codesystems) != length(searchvars)) {
    stop("The number of code systems and search-variable groups must be equal.",
         call. = FALSE)
  }
  if (length(codesystems) == 0L || any(lengths(searchvars) == 0L)) {
    stop("codesystems and searchvars must not be empty.", call. = FALSE)
  }

  canonical <- vapply(codesystems, .impact_normalize_codesystem, character(1))
  for (cols in searchvars) {
    missing_cols <- setdiff(cols, names(data))
    if (length(missing_cols)) {
      stop("Search variable(s) not found: ", paste(missing_cols, collapse = ", "),
           call. = FALSE)
    }
    non_character <- cols[!vapply(data[cols], is.character, logical(1))]
    if (length(non_character)) {
      stop(
        "Search variable(s) must be character columns: ",
        paste(non_character, collapse = ", "),
        ". Store clinical codes as character values to preserve leading ",
        "zeroes and long identifiers.",
        call. = FALSE
      )
    }
  }

  target_ids <- if (level == "ltc") .impact_ltc_ids else .impact_phenotype_ids
  indicator_names <- paste0("__", target_ids)
  flags <- matrix(0L, nrow = nrow(data), ncol = length(target_ids),
                  dimnames = list(NULL, indicator_names))
  result <- cbind(data[id], as.data.frame(flags, check.names = FALSE))

  lookups <- vector("list", length(canonical))
  for (i in seq_along(canonical)) {
    lookup <- get(canonical[[i]], envir = .impact_lookup_store, inherits = FALSE)
    lookups[[i]] <- lookup
    for (column in searchvars[[i]]) {
      values <- trimws(as.character(data[[column]]))
      for (code in unique(values[!is.na(values) & values != ""])) {
        ltcs <- lookup[[code]]
        if (is.null(ltcs)) next
        targets <- if (level == "ltc") {
          unique(as.character(ltcs))
        } else {
          unique(unname(.impact_ltc_phenotypes[as.character(ltcs)]))
        }
        rows <- which(!is.na(values) & values == code)
        result[rows, paste0("__", targets)] <- 1L
      }
    }
  }

  if (isTRUE(multimorbidity)) {
    result <- .impact_add_multimorbidity(result, level)
  }

  if (isTRUE(summary)) {
    cat("\nIMPACT summary:", nrow(data), "coded event(s)\n")
    for (i in seq_along(canonical)) {
      keys <- ls(lookups[[i]], all.names = TRUE)
      total <- sum(vapply(searchvars[[i]], function(column) {
        values <- trimws(as.character(data[[column]]))
        sum(!is.na(values) & values %in% keys)
      }, integer(1)))
      cat(" ", codesystems[[i]], ": ", total, " matched code(s)\n", sep = "")
    }
  }

  result
}

.impact_validate_logical <- function(value, name) {
  if (!is.logical(value) || length(value) != 1L || is.na(value)) {
    stop(name, " must be either TRUE or FALSE.", call. = FALSE)
  }
}

.impact_as_codesystems <- function(value) {
  if (is.character(value) && length(value) == 1L) {
    value <- strsplit(value, "[,[:space:]]+")[[1L]]
  }
  value <- tolower(as.character(unlist(value, use.names = FALSE)))
  value[nzchar(value)]
}

.impact_canonical_codesystems <- c(
  "cprd_aurum_medcodeid", "cprd_gold_medcode", "emis_local", "icd10",
  "icd10cm", "icd10pcs", "icd9cm", "icd9pcs", "opcs4",
  "read_cleansed", "read_original", "snomed_concept", "snomed_description"
)

.impact_normalize_codesystem <- function(value) {
  value <- tolower(value)
  if (!(value %in% .impact_canonical_codesystems)) {
    stop("Unknown code system '", value, "'. See list_codesystems().", call. = FALSE)
  }
  value
}

.impact_system_name <- function(value) {
  value <- tolower(gsub("&", "and", value, fixed = TRUE))
  value <- gsub("[^a-z0-9]+", "_", value)
  sub("_$", "", sub("^_", "", value))
}

.impact_add_multimorbidity <- function(result, level) {
  phenotype_names <- paste0("__", .impact_phenotype_ids)
  if (level == "phenotype") {
    phenotype_flags <- as.matrix(result[phenotype_names])
  } else {
    phenotype_flags <- matrix(
      0L, nrow = nrow(result), ncol = length(.impact_phenotype_ids),
      dimnames = list(NULL, phenotype_names)
    )
    for (ltc in .impact_ltc_ids) {
      phenotype <- unname(.impact_ltc_phenotypes[[ltc]])
      column <- paste0("__", phenotype)
      phenotype_flags[, column] <- pmax(
        phenotype_flags[, column], result[[paste0("__", ltc)]]
      )
    }
  }

  result[["__nphenotypes"]] <- rowSums(phenotype_flags)
  mental <- .impact_phenotype_ids[.impact_phenotype_categories == "Mental"]
  result[["__nmental"]] <- rowSums(phenotype_flags[, paste0("__", mental), drop = FALSE])
  result[["__nphysical"]] <- result[["__nphenotypes"]] - result[["__nmental"]]

  systems <- unique(unname(.impact_phenotype_systems))
  body_present <- matrix(FALSE, nrow = nrow(result), ncol = length(systems))
  for (i in seq_along(systems)) {
    phenotypes <- .impact_phenotype_ids[.impact_phenotype_systems == systems[[i]]]
    counts <- rowSums(phenotype_flags[, paste0("__", phenotypes), drop = FALSE])
    result[[paste0("__bs_", .impact_system_name(systems[[i]]))]] <- counts
    body_present[, i] <- counts > 0L
  }
  result[["__nbody"]] <- rowSums(body_present)
  result
}

#' Return an LTC lookup for a supported code system.
#'
#' @param ontology A value returned by `list_codesystems()`.
#' @return An independent environment mapping code strings to LTC identifier
#'   vectors. Modifying it does not alter IMPACT's internal lookup.
#' @export
select_codesystem <- function(ontology) {
  canonical <- .impact_normalize_codesystem(ontology)
  lookup <- get(canonical, envir = .impact_lookup_store, inherits = FALSE)
  list2env(
    as.list.environment(lookup, all.names = TRUE),
    hash = TRUE,
    parent = emptyenv()
  )
}

#' List supported code-system names.
#'
#' @return A character vector.
#' @export
list_codesystems <- function() {
  .impact_canonical_codesystems
}

#' List IMPACT granular LTC metadata.
#'
#' @return A data frame with LTC and grouped-phenotype metadata.
#' @export
list_ltcs <- function() {
  phenotypes <- unname(.impact_ltc_phenotypes[.impact_ltc_ids])
  data.frame(
    ltc = .impact_ltc_ids,
    label = unname(.impact_ltc_labels[.impact_ltc_ids]),
    phenotype = phenotypes,
    phenotype_label = unname(.impact_phenotype_labels[phenotypes]),
    body_system = unname(.impact_phenotype_systems[phenotypes]),
    category = unname(.impact_phenotype_categories[phenotypes]),
    stringsAsFactors = FALSE
  )
}

File: packages/python/src/impact/impact.py
"""Inclusive Multimorbidity Phenotyping Algorithm and Coding Tool (IMPACT)."""

import gzip
import json
from functools import lru_cache
import pkgutil
import re

__all__ = ["impact", "select_codesystem", "list_ltcs", "list_codesystems"]


def _load_json_resource(filename):
    raw = pkgutil.get_data(__package__, "data/" + filename)
    if raw is None:
        raise RuntimeError("IMPACT package resource %r is missing." % filename)
    try:
        payload = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise RuntimeError(
            "IMPACT package resource %r is not valid UTF-8 JSON." % filename
        ) from exc
    if payload.get("schema_version") != 1:
        raise RuntimeError(
            "Unsupported schema in IMPACT package resource %r." % filename
        )
    return payload


_LTC_RECORDS = _load_json_resource("ltcs.json")["ltcs"]
_PHENOTYPE_RECORDS = _load_json_resource("phenotypes.json")["phenotypes"]

LTC_IDS = tuple(row["ltc_id"] for row in _LTC_RECORDS)
LTC_LABELS = {row["ltc_id"]: row["ltc_label"] for row in _LTC_RECORDS}
LTC_PHENOTYPES = {
    row["ltc_id"]: row["phenotype_id"] for row in _LTC_RECORDS
}
PHENOTYPE_IDS = tuple(row["phenotype_id"] for row in _PHENOTYPE_RECORDS)
PHENOTYPE_LABELS = {
    row["phenotype_id"]: row["phenotype_label"]
    for row in _PHENOTYPE_RECORDS
}
PHENOTYPE_SYSTEMS = {
    row["phenotype_id"]: row["body_system"] for row in _PHENOTYPE_RECORDS
}
PHENOTYPE_CATEGORIES = {
    row["phenotype_id"]: row["category"] for row in _PHENOTYPE_RECORDS
}

CANONICAL_CODESYSTEMS = (
    "cprd_aurum_medcodeid",
    "cprd_gold_medcode",
    "emis_local",
    "icd10",
    "icd10cm",
    "icd10pcs",
    "icd9cm",
    "icd9pcs",
    "opcs4",
    "read_cleansed",
    "read_original",
    "snomed_concept",
    "snomed_description",
)

ALIASES = {
    "cprdaurum": "cprd_aurum_medcodeid",
    "medcodeid": "cprd_aurum_medcodeid",
    "cprdgold": "cprd_gold_medcode",
}


def _normalise_codesystem(value):
    value = str(value).lower()
    value = ALIASES.get(value, value)
    if value not in CANONICAL_CODESYSTEMS:
        raise ValueError(
            "Unknown code system %r. See list_codesystems()." % value
        )
    return value


@lru_cache(maxsize=1)
def _load_all_codes():
    raw = pkgutil.get_data(__package__, "data/codes.json.gz")
    if raw is None:
        raise RuntimeError("IMPACT package resource 'codes.json.gz' is missing.")
    try:
        payload = json.loads(gzip.decompress(raw).decode("utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise RuntimeError(
            "IMPACT package resource 'codes.json.gz' is not valid UTF-8 JSON."
        ) from exc
    if payload.get("schema_version") != 1:
        raise RuntimeError(
            "Unsupported schema in IMPACT package resource 'codes.json.gz'."
        )
    return payload["codesystems"]


def _load_codesystem(canonical):
    """Load a generated code-to-LTC mapping from package resources."""
    try:
        return _load_all_codes()[canonical]
    except KeyError as exc:
        raise RuntimeError(
            "Code system %r is absent from IMPACT package resources." % canonical
        ) from exc


def select_codesystem(ontology):
    """Return the code-to-granular-LTC dictionary for one code system."""
    return _load_codesystem(_normalise_codesystem(ontology))


def list_codesystems():
    """Return canonical supported code-system names followed by aliases."""
    return list(CANONICAL_CODESYSTEMS) + list(ALIASES)


def list_ltcs():
    """Return granular LTC metadata and each LTC's grouped phenotype."""
    import pandas as pd

    phenotypes = [LTC_PHENOTYPES[x] for x in LTC_IDS]
    return pd.DataFrame(
        {
            "ltc": LTC_IDS,
            "label": [LTC_LABELS[x] for x in LTC_IDS],
            "phenotype": phenotypes,
            "phenotype_label": [PHENOTYPE_LABELS[x] for x in phenotypes],
            "body_system": [PHENOTYPE_SYSTEMS[x] for x in phenotypes],
            "category": [PHENOTYPE_CATEGORIES[x] for x in phenotypes],
        }
    )


def _as_codesystems(value):
    if isinstance(value, str):
        values = [x for x in re.split(r"[,\s]+", value) if x]
    else:
        values = list(value)
    if not values:
        raise ValueError("codesystems must not be empty.")
    return [_normalise_codesystem(x) for x in values]


def _as_search_groups(value, number_of_systems):
    if isinstance(value, str):
        return [[value]]
    values = list(value)
    if number_of_systems == 1 and all(isinstance(x, str) for x in values):
        return [values]
    groups = [[x] if isinstance(x, str) else list(x) for x in values]
    return groups


def impact(
    df,
    id,
    codesystems,
    searchvars,
    level,
    multimorbidity=False,
    summary=False,
):
    """Ascertain IMPACT indicators from a pandas DataFrame of coded events.

    Parameters
    ----------
    df : pandas.DataFrame
        One row per coded event.
    id : str
        Identifier column retained in the result.
    codesystems : str or sequence of str
        Code systems corresponding, in order, to ``searchvars``.
    searchvars : str or sequence
        One code-column group per code system. A nested sequence searches
        several columns using one code system.
    level : {"ltc", "phenotype"}
        Return 321 granular LTC indicators or 116 grouped phenotype indicators.
    multimorbidity : bool, optional
        Add phenotype-based counts, including ``__nphenotypes``.
    summary : bool, optional
        Print matched-code counts for each selected code system.
    """
    import pandas as pd

    if not isinstance(df, pd.DataFrame):
        raise TypeError("df must be a pandas DataFrame.")
    if not isinstance(id, str) or not id:
        raise TypeError("id must be one column name.")
    if id not in df.columns:
        raise KeyError("Identifier column %r was not found." % id)
    if not isinstance(level, str) or level.lower() not in ("ltc", "phenotype"):
        raise ValueError("level must be specified as 'ltc' or 'phenotype'.")
    level = level.lower()
    canonical = _as_codesystems(codesystems)
    search_groups = _as_search_groups(searchvars, len(canonical))
    if len(canonical) != len(search_groups):
        raise ValueError(
            "The number of code systems and search-variable groups must be equal."
        )
    if any(not group for group in search_groups):
        raise ValueError("searchvars must not contain an empty group.")
    for group in search_groups:
        missing = [column for column in group if column not in df.columns]
        if missing:
            raise KeyError("Search variable(s) not found: %s" % ", ".join(missing))

    target_ids = LTC_IDS if level == "ltc" else PHENOTYPE_IDS
    flags = pd.DataFrame(
        0,
        index=df.index,
        columns=["__" + x for x in target_ids],
        dtype="int8",
    )
    result = pd.concat([df[[id]].copy(), flags], axis=1)

    lookups = []
    for canonical_name, columns in zip(canonical, search_groups):
        lookup = _load_codesystem(canonical_name)
        lookups.append(lookup)
        for column in columns:
            values = df[column].astype("string").str.strip()
            for code in values.dropna().unique():
                if code == "":
                    continue
                ltcs = lookup.get(str(code))
                if not ltcs:
                    continue
                if isinstance(ltcs, str):
                    ltcs = [ltcs]
                targets = (
                    list(dict.fromkeys(ltcs))
                    if level == "ltc"
                    else list(dict.fromkeys(LTC_PHENOTYPES[x] for x in ltcs))
                )
                result.loc[values == code, ["__" + x for x in targets]] = 1

    if multimorbidity:
        _add_multimorbidity(result, level)

    if summary:
        print("\nIMPACT summary: %d coded event(s)" % len(df))
        for requested, columns, lookup in zip(
            _display_codesystems(codesystems), search_groups, lookups
        ):
            total = 0
            for column in columns:
                values = df[column].astype("string").str.strip()
                total += int(values.isin(lookup).sum())
            print("  %s: %d matched code(s)" % (requested, total))

    return result


def _display_codesystems(value):
    if isinstance(value, str):
        return [x for x in re.split(r"[,\s]+", value) if x]
    return [str(x) for x in value]


def _system_name(value):
    value = value.lower().replace("&", "and")
    return re.sub(r"^_|_$", "", re.sub(r"[^a-z0-9]+", "_", value))


def _add_multimorbidity(result, level):
    import pandas as pd

    phenotype_columns = ["__" + x for x in PHENOTYPE_IDS]
    if level == "phenotype":
        phenotype_flags = result[phenotype_columns].copy()
    else:
        phenotype_flags = pd.DataFrame(
            0, index=result.index, columns=phenotype_columns, dtype="int8"
        )
        for ltc in LTC_IDS:
            column = "__" + LTC_PHENOTYPES[ltc]
            phenotype_flags[column] = phenotype_flags[column].where(
                result["__" + ltc] == 0, 1
            )

    result["__nphenotypes"] = phenotype_flags.sum(axis=1).astype("int16")
    mental = [
        "__" + x for x in PHENOTYPE_IDS if PHENOTYPE_CATEGORIES[x] == "Mental"
    ]
    result["__nmental"] = phenotype_flags[mental].sum(axis=1).astype("int16")
    result["__nphysical"] = result["__nphenotypes"] - result["__nmental"]

    systems = list(dict.fromkeys(PHENOTYPE_SYSTEMS[x] for x in PHENOTYPE_IDS))
    body_present = pd.DataFrame(False, index=result.index, columns=systems)
    for system in systems:
        columns = [
            "__" + x for x in PHENOTYPE_IDS if PHENOTYPE_SYSTEMS[x] == system
        ]
        counts = phenotype_flags[columns].sum(axis=1).astype("int16")
        result["__bs_" + _system_name(system)] = counts
        body_present[system] = counts > 0
    result["__nbody"] = body_present.sum(axis=1).astype("int16")
